### PR TITLE
fix(qwen35): batch linear decode kernels

### LINKED ISSUE #689: [qwen35] Attribute and reduce HTTP c16 decode graph GPU inflation
## Summary

Qwen3.5-4B on a single RTX 5090 has a retained #469 HTTP serving gap at `1024/256 c16`. The best current attribution points to Qwen3.5 decode CUDA graph cost at HTTP bucket 16 / higher load, with frontend serialization and output transport ruled out as primary causes.

## Evidence

Environment boundary from the retained attribution run:

- GPU: 1x NVIDIA GeForce RTX 5090 32GB
- OpenInfer source boundary: `e2de05dbc52bedbbb3c213e648148b64cd12b3b4`
- Model: `Qwen/Qwen3.5-4B`
- Client: `vllm bench serve`, prefix cache disabled for the primary comparison
- Workload focus: random `input_len=1024`, `output_len=256`, `max_concurrency=16`, greedy, ignore-eos

Measured result:

- OpenInfer HTTP c16: TPOT about `17.35ms`, output about `804 tok/s`, zero failed requests in the retained runs.
- vLLM 0.25.1 HTTP c16: TPOT about `9.34ms`, output about `1425 tok/s`.
- OpenInfer HTTP c16 trace: scheduler is mostly Decode ticks, and the decode CUDA graph event time averages about `15.15ms`, roughly `87%` of TPOT.
- Bridge/msgpack/ZMQ/output loop costs are microseconds-scale, so they are not the primary gap.
- Higher request-rate pressure increases graph GPU elapsed further: QPS8 about `13.97ms`, QPS12 about `19.21ms`, QPS16 about `20.16ms`.

Important attribution correction: the earlier direct diagnostic labelled c16 used the direct `bench_serving` scheduler path with Qwen3.5 max batch hardcoded to `4`, so its `9.16ms` TPOT / `8.72ms` graph event time is useful as a bs4 diagnostic only. It should not be read as evidence that the bucket16 graph is already near vLLM.

## Current hypothesis

The next optimization target should be the Qwen3.5 bucket16 decode graph / kernels under HTTP serving shape, not bridge batching or output transport. Kernel profiling under HTTP c16 or an equivalent true bucket16 probe is needed before changing the serving claim.

## Required acceptance evidence

A fix should include a retained artifact bundle with:

- Exact OpenInfer serve command/env and exact benchmark commands.
- Same host / same model / same client before-and-after comparison.
- HTTP `1024/256` c1 repeated 3x.
- HTTP `1024/256` c16 repeated 3x.
- QPS 8 / 12 / 16 if the run budget allows.
- For every cell: completed, failed, timeout/unsupported/OOM if applicable, average output tokens, output sanity/hash, TTFT, TPOT, ITL, output tok/s.
- Trace summary for scheduler plan mix, decode batch/bucket distribution, batch_decode_graph wall time, CUDA event graph elapsed, metadata sync, sampling, token dispatch, and frontend output-loop cost.
- Server stderr warnings/errors.

Direct in-process diagnostics may be used for attribution only when the actual graph batch/bucket distribution is recorded.

## Non-goals

- No prefix-cache product-mode comparison in the primary result.
- No vLLM parity, SOTA, or production-ready claim unless separately proven by a fair retained benchmark matrix.
- No dropping failed/timeout/unsupported/OOM cells from the report.

### PR DESCRIPTION
**Pull request by CAICAIIs**
## Summary

- batch Qwen3.5 linear-attention decode conv1d/GDR across active decode slots instead of launching per-slot kernels inside the decode graph
- keep CUDA Graph replay address-stable by precomputing per-linear-layer device pointer tables for recurrent and conv state slots
- remove obsolete per-slot decode scratch buffers from the batch decode hot path

Fixes #689.

## Evidence

Artifact bundle is retained on the RTX 5090 validation host. Public summary paths are intentionally omitted from the PR body; the bundle contains raw benchmark JSON/stdout, server logs, trace JSONL, nsys report/sqlite, parsed CSV/JSON, and README.

Validation:

```bash
# RTX 5090 validation host. Replace paths with the local equivalents.
export OPENINFER_CUDA_SM=120
export OPENINFER_TRITON_PYTHON=<python-with-triton>
export CUDA_HOME=<cuda-toolkit-root>
export OPENINFER_TEST_MODEL_PATH=<path-to-Qwen3.5-4B-weights>

cargo fmt --check

cargo build --release -p openinfer-server --features qwen35-4b

cargo build --release --bin bench_serving --features qwen35-4b

cargo test --release -p openinfer-qwen35-4b --features qwen35-4b --test e2e_scheduler -- --nocapture

cargo test --release -p openinfer-qwen35-4b --features qwen35-4b --test hf_golden_gate -- --nocapture
```

Additional post-review guard validation:

```bash
cargo test --release -p openinfer-qwen35-4b --features qwen35-4b config::tests -- --nocapture
```

HTTP serving benchmark, prefix cache not enabled, same `vllm bench serve` client:

| Cell | Runs | completed | failed | avg out | mean TPOT | output tok/s | graph GPU |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1024/256 c1 | 3 | 4/run | 0 | 256.0 | 6.90ms | 141.2 | 6.85ms |
| 1024/256 c16 | 3 | 16/run | 0 | 256.0 | 11.30ms | 1173.5 | 9.58ms |
| 1024/256 QPS 8 | 1 | 16 | 0 | 256.0 | 10.78ms | 906.5 | 8.91ms |
| 1024/256 QPS 12 | 1 | 24 | 0 | 256.0 | 11.85ms | 1371.4 | 9.05ms |
| 1024/256 QPS 16 | 1 | 32 | 0 | 256.0 | 12.81ms | 1759.4 | 9.37ms |

Delta vs retained baseline artifact:

- HTTP `1024/256 c16` mean TPOT: `17.35ms -> 11.30ms` (`-34.8%`)
- HTTP `1024/256 c16` decode graph GPU elapsed: `15.15ms -> 9.58ms` (`-36.8%`)
- HTTP `1024/256 c16` output throughput: `804.0 -> 1173.5 tok/s` (`+46.0%`)

An additional `nsys --cuda-graph-trace=node` composition check on HTTP c16/output64 confirms the expanded graph now contains:

- `gated_delta_rule_decode_batch_kernel`
- `conv1d_decode_batch_kernel`

The nsys run is not used for TPOT claims because profiler overhead inflates latency.

## Claim boundary

This PR does not claim vLLM parity, SOTA, or production readiness. It only claims the measured OpenInfer HTTP c16 regression bucket is reduced on the retained single RTX 5090 setup, with raw artifacts preserved.

### CODE REVIEW

**Comment by xiaguan** [comment]
Thanks — the direction here is exactly right, and the win is real. I read the new kernels line-by-line against the single-slot ones: the math matches, including the bf16-round-before-SiLU detail in conv1d. Nice work keeping that faithful.

A few things before this can land:

1. **The head commit doesn't build with `--features qwen35-4b`** — `batch_decode.rs` uses `RecurrentState` but the `use super::recurrent_state::RecurrentState;` import is gone (E0425, two sites). CI is green only because the workflow never compiles this feature. I suspect the branch that ran your validation bundle and the branch that got pushed are not quite the same tree — could you push the one the evidence came from?

2. **`linear_state_pointer_tables` runs on the TP eager path per decode step** (`tp_executor` → `batch_decode_eager_logits`), which means 24 layers × 2 `clone_htod` allocations per token — for pointers that never change during a request's lifetime. The graph path caches its tables at init; the eager path deserves the same treatment. As it stands, the freshly-landed TP path quietly picks up a per-token driver-call tab that nobody has measured yet.

3. Minor: the pointer-table builder loop now exists twice (`batch_decode.rs` and `BatchDecodeGraphState::new`) with only the error strings differing — worth extracting a shared helper. And since `gated_delta_rule_decode_vec_into` is already retained as a `#[cfg(test)]` reference, a small batch-vs-single equivalence test at bs>1 would be cheap insurance.

FWIW I ran this locally on a 16GB RTX 5070 Ti (with the one-line import fix): `hf_golden_gate` passes all four modes (sequential bs=1, batched 5/3-padded, slot compaction) and `e2e_scheduler` phases 1–5 pass, so the kernels themselves check out.

**Comment by CAICAIIs** [comment]
> Thanks — the direction here is exactly right, and the win is real. I read the new kernels line-by-line against the single-slot ones: the math matches, including the bf16-round-before-SiLU detail in conv1d. Nice work keeping that faithful.
> 
> A few things before this can land:
> 
> 1. **The head commit doesn't build with `--features qwen35-4b`** — `batch_decode.rs` uses `RecurrentState` but the `use super::recurrent_state::RecurrentState;` import is gone (E0425, two sites). CI is green only because the workflow never compiles this feature. I suspect the branch that ran your validation bundle and the branch that got pushed are not quite the same tree — could you push the one the evidence came from?
> 2. **`linear_state_pointer_tables` runs on the TP eager path per decode step** (`tp_executor` → `batch_decode_eager_logits`), which means 24 layers × 2 `clone_htod` allocations per token — for pointers that never change during a request's lifetime. The graph path caches its tables at init; the eager path deserves the same treatment. As it stands, the freshly-landed TP path quietly picks up a per-token driver-call tab that nobody has measured yet.
> 3. Minor: the pointer-table builder loop now exists twice (`batch_decode.rs` and `BatchDecodeGraphState::new`) with only the error strings differing — worth extracting a shared helper. And since `gated_delta_rule_decode_vec_into` is already retained as a `#[cfg(test)]` reference, a small batch-vs-single equivalence test at bs>1 would be cheap insurance.
> 
> FWIW I ran this locally on a 16GB RTX 5070 Ti (with the one-line import fix): `hf_golden_gate` passes all four modes (sequential bs=1, batched 5/3-padded, slot compaction) and `e2e_scheduler` phases 1–5 pass, so the kernels themselves check out.

done, thanks your review

**Comment by xiaguan** [comment]
@codex review
